apps/remote_control/lib/lexical/remote_control/project_node.ex

```diff
@@ -0,0 +1,210 @@
+defmodule Lexical.RemoteControl.ProjectNode do
+  alias Lexical.Project
+  alias Lexical.RemoteControl
+  require Logger
+
+  defmodule State do
+    defstruct [
+      :project,
+      :port,
+      :cookie,
+      :stopped_by,
+      :stop_timeout,
+      :started_by,
+      :status
+    ]
+
+    def new(%Project{} = project) do
+      cookie = Node.get_cookie()
+
+      %__MODULE__{
+        project: project,
+        cookie: cookie,
+        status: :initializing
+      }
+    end
+
+    @dialyzer {:nowarn_function, start: 3}
+
+    def start(%__MODULE__{} = state, paths, from) do
+      port_wrapper = port_wrapper_executable()
+      {:ok, elixir_executable} = RemoteControl.elixir_executable(state.project)
+      node_name = node_name(state.project)
+
+      args = [
+        elixir_executable,
+        "--name",
+        node_name,
+        "--cookie",
+        state.cookie,
+        "--no-halt",
+        "-e",
+        "Node.connect(#{inspect(Node.self())})"
+        | path_append_arguments(paths)
+      ]
+
+      port =
+        Port.open({:spawn_executable, port_wrapper},
+          args: args,
+          cd: Project.root_path(state.project)
+        )
+
+      %{state | port: port, started_by: from}
+    end
+
+    def stop(%__MODULE__{} = state, from, stop_timeout) do
+      :rpc.call(node_name(state.project), System, :stop, [])
+      %{state | stopped_by: from, stop_timeout: stop_timeout, status: :stopping}
+    end
+
+    def halt(%__MODULE__{} = state) do
+      :rpc.call(node_name(state.project), System, :halt, [])
+      on_nodedown(state)
+    end
+
+    def handle_nodeup(%__MODULE__{} = state) do
+      %{state | status: :started}
+    end
+
+    def on_nodedown(%__MODULE__{} = state) do
+      %{state | status: :stopped}
+    end
+
+    def node_name(%Project{} = project) do
+      :"#{Project.name(project)}@127.0.0.1"
+    end
+
+    defp path_append_arguments(paths) do
+      Enum.flat_map(paths, fn path ->
+        ["-pa", Path.expand(path)]
+      end)
+    end
+
+    defp port_wrapper_executable do
+      :remote_control
+      |> :code.priv_dir()
+      |> Path.join("port_wrapper.sh")
+    end
+  end
+
+  alias Lexical.RemoteControl.ProjectNodeSupervisor
+  use GenServer
+
+  def start(project, project_listener, paths) do
+    node_name = State.node_name(project)
+    remote_control_config = Application.get_all_env(:remote_control)
+
+    {:ok, node_pid} = ProjectNodeSupervisor.start_project_node(project)
```

Comment:
Should this go into the with? Do you want to crash if this fails?

Reply:
Should rename this, because     
```elixir
with {:ok, node_pid} <- ProjectNodeSupervisor.start_project_node(project),
         :ok <- start_node(project, paths),
 ```
` ProjectNodeSupervisor.start_project_node` followed by a `start_node` looks not good